File: sdk/runanywhere-swift/Modules/ONNXRuntime/Sources/ONNXRuntime/ONNXAdapter.swift
import Foundation
import RunAnywhere

/// ONNX Runtime adapter for multi-modal inference
public class ONNXAdapter: UnifiedFrameworkAdapter {
    private let logger = SDKLogger(category: "ONNXAdapter")

    // Singleton instance to ensure caching works across the app
    public static let shared = ONNXAdapter()

    // MARK: - Properties

    public let framework: LLMFramework = .onnx

    public let supportedModalities: Set<FrameworkModality> = [.voiceToText, .textToText]

    public let supportedFormats: [ModelFormat] = [.onnx, .ort]

    // Cache service instances to avoid re-initialization
    private var cachedSTTService: ONNXSTTService?
    private var cachedLLMService: Any? // Reserved for future LLM implementation

    // Track last usage for smart cleanup
    private var lastSTTUsage: Date?
    private let cacheTimeout: TimeInterval = 300 // 5 minutes

    // MARK: - UnifiedFrameworkAdapter Implementation

    public func canHandle(model: ModelInfo) -> Bool {
        // Check if model is compatible with ONNX Runtime
        let canHandle = model.compatibleFrameworks.contains(.onnx) &&
                       (model.category == .speechRecognition || model.category == .language)
        logger.debug("canHandle(\(model.name)): \(canHandle)")
        return canHandle
    }

    public func createService(for modality: FrameworkModality) -> Any? {
        logger.info("createService for modality: \(modality.rawValue)")
        switch modality {
        case .voiceToText:
            // Check if cached service should be cleaned up
            cleanupStaleCache()

            // Return cached instance if available
            if let cached = cachedSTTService {
                logger.info("Returning cached ONNXSTTService for voice-to-text")
                lastSTTUsage = Date()
                return cached
            }
            logger.info("Creating new ONNXSTTService for voice-to-text")
            let service = ONNXSTTService()
            cachedSTTService = service
            lastSTTUsage = Date()
            return service

        case .textToText:
            // Reserved for future LLM implementation
            logger.warning("LLM support not yet implemented")
            return nil

        default:
            logger.warning("Unsupported modality: \(modality.rawValue)")
            return nil
        }
    }

    public func loadModel(_ model: ModelInfo, for modality: FrameworkModality) async throws -> Any {
        logger.info("loadModel(\(model.name)) for modality: \(modality.rawValue)")
        switch modality {
        case .voiceToText:
            // Check if cached service should be cleaned up
            cleanupStaleCache()

            // Use cached service if available
            let service: ONNXSTTService
            if let cached = cachedSTTService {
                logger.info("Using cached ONNXSTTService for initialization")
                service = cached
            } else {
                logger.info("Creating new ONNXSTTService for initialization")
                service = ONNXSTTService()
                cachedSTTService = service
            }

            // Initialize with model path if available
            let modelPath = model.localPath?.path
            logger.debug("Model path: \(modelPath ?? "nil")")
            try await service.initialize(modelPath: modelPath)
            logger.info("ONNXSTTService initialized")
            lastSTTUsage = Date()
            return service

        case .textToText:
            logger.error("LLM support not yet implemented")
            throw SDKError.unsupportedModality(modality.rawValue)

        default:
            logger.error("Unsupported modality: \(modality.rawValue)")
            throw SDKError.unsupportedModality(modality.rawValue)
        }
    }

    public func configure(with hardware: HardwareConfiguration) async {
        // ONNX Runtime will use CPU by default
        // Future: Could configure execution providers here (CoreML, etc.)
        logger.info("Hardware configuration: \(hardware.primaryAccelerator.rawValue)")
    }

    public func estimateMemoryUsage(for model: ModelInfo) -> Int64 {
        return model.memoryRequired ?? 0
    }

    public func optimalConfiguration(for model: ModelInfo) -> HardwareConfiguration {
        // For now, use CPU. Future: detect available hardware and use CoreML if available
        return HardwareConfiguration(primaryAccelerator: .cpu)
    }

    /// Called when adapter is registered with the SDK
    /// Registers the STT service provider with ModuleRegistry
    @MainActor
    public func onRegistration() {
        ModuleRegistry.shared.registerSTT(ONNXSTTServiceProvider())
        logger.info("Registered ONNXServiceProvider with ModuleRegistry")
    }

    public func getProvidedModels() -> [ModelInfo] {
        // No pre-bundled models
        return []
    }

    public func getDownloadStrategy() -> DownloadStrategy? {
        // Return custom strategy for handling .tar.bz2 archives
        return ONNXDownloadStrategy()
    }

    public func initializeComponent(
        with parameters: any ComponentInitParameters,
        for modality: FrameworkModality
    ) async throws -> Any? {
        // Use createService for initialization
        return createService(for: modality)
    }

    // MARK: - Private Helpers

    private func cleanupStaleCache() {
        // Clean up STT service if not used recently
        if let lastUsage = lastSTTUsage,
           Date().timeIntervalSince(lastUsage) > cacheTimeout {
            logger.info("Cleaning up stale ONNXSTTService cache")
            cachedSTTService = nil
            lastSTTUsage = nil
        }
    }
}

File: sdk/runanywhere-kotlin/modules/runanywhere-core-onnx/src/commonMain/kotlin/com/runanywhere/sdk/core/onnx/ONNXAdapter.kt
package com.runanywhere.sdk.core.onnx

import com.runanywhere.sdk.features.llm.HardwareConfiguration
import com.runanywhere.sdk.core.ModuleRegistry
import com.runanywhere.sdk.core.frameworks.ComponentInitParameters
import com.runanywhere.sdk.core.frameworks.DownloadStrategy
import com.runanywhere.sdk.core.frameworks.ModelStorageStrategy
import com.runanywhere.sdk.core.frameworks.UnifiedFrameworkAdapter
import com.runanywhere.sdk.foundation.SDKLogger
import com.runanywhere.sdk.foundation.currentTimeMillis
import com.runanywhere.sdk.models.ModelInfo
import com.runanywhere.sdk.models.enums.FrameworkModality
import com.runanywhere.sdk.models.enums.InferenceFramework
import com.runanywhere.sdk.models.enums.ModelFormat

/**
 * ONNX Runtime Framework Adapter
 *
 * Unified adapter for ONNX Runtime, supporting multiple modalities:
 * - Speech-to-Text (Whisper, Zipformer, Paraformer)
 * - Text-to-Speech (Piper, VITS)
 * - Voice Activity Detection (Silero VAD)
 * - Text Generation (future)
 *
 * Matches iOS ONNXAdapter
 * Reference: sdk/runanywhere-swift/Sources/ONNXRuntime/ONNXAdapter.swift
 */
class ONNXAdapter : UnifiedFrameworkAdapter {

    private val logger = SDKLogger("ONNXAdapter")

    // Cached service instances for reuse
    private var cachedSTTService: Any? = null
    private var cachedTTSService: Any? = null
    private var lastSTTUsage: Long? = null
    private var lastTTSUsage: Long? = null
    private val cacheTimeout: Long = 300_000 // 5 minutes in milliseconds

    // MARK: - UnifiedFrameworkAdapter Properties

    override val framework: InferenceFramework = InferenceFramework.ONNX

    override val supportedModalities: Set<FrameworkModality> = setOf(
        FrameworkModality.VOICE_TO_TEXT,
        FrameworkModality.TEXT_TO_VOICE
        // Note: TEXT_TO_TEXT not yet implemented - removed to avoid contract violation
    )

    override val supportedFormats: List<ModelFormat> = listOf(
        ModelFormat.ONNX,
        ModelFormat.ORT
    )

    // MARK: - UnifiedFrameworkAdapter Methods

    /**
     * Check if this adapter can handle the given model
     * Matches iOS canHandle(model:) implementation
     */
    override fun canHandle(model: ModelInfo): Boolean {
        // Check format support
        if (!supportedFormats.contains(model.format)) {
            return false
        }

        // Check framework compatibility
        if (model.compatibleFrameworks.isNotEmpty()) {
            return model.compatibleFrameworks.contains(InferenceFramework.ONNX)
        }

        // Check by model ID patterns
        val modelId = model.id.lowercase()
        return modelId.contains("onnx") ||
                modelId.contains("zipformer") ||
                modelId.contains("sherpa") ||
                modelId.contains("piper") ||
                modelId.contains("vits") ||
                modelId.contains("silero")
    }

    /**
     * Create a service instance for the given modality
     */
    override fun createService(modality: FrameworkModality): Any? {
        cleanupStaleCache()

        return when (modality) {
            FrameworkModality.VOICE_TO_TEXT -> {
                lastSTTUsage = currentTimeMillis()
                cachedSTTService
            }
            FrameworkModality.TEXT_TO_VOICE -> {
                lastTTSUsage = currentTimeMillis()
                cachedTTSService
            }
            else -> {
                logger.warning("Modality $modality not supported by ONNX adapter")
                null
            }
        }
    }

    /**
     * Load a model and return a service instance
     */
    override suspend fun loadModel(model: ModelInfo, modality: FrameworkModality): Any {
        logger.info("Loading model ${model.id} for modality $modality")

        val localPath = model.localPath
            ?: throw ONNXError.ModelNotFound(model.id)

        return when (modality) {
            FrameworkModality.VOICE_TO_TEXT -> {
                val service = createONNXSTTServiceFromPath(localPath)
                cachedSTTService = service
                lastSTTUsage = currentTimeMillis()
                service
            }
            FrameworkModality.TEXT_TO_VOICE -> {
                val service = createONNXTTSServiceFromPath(localPath)
                cachedTTSService = service
                lastTTSUsage = currentTimeMillis()
                service
            }
            else -> {
                throw ONNXError.NotImplemented
            }
        }
    }

    /**
     * Configure adapter with hardware settings
     */
    override suspend fun configure(hardware: HardwareConfiguration) {
        logger.debug("Configuring ONNX adapter with hardware: $hardware")
        // ONNX Runtime configuration is handled at service level
    }

    /**
     * Estimate memory usage for a model
     * Matches iOS estimateMemoryUsage(for:) implementation
     */
    override fun estimateMemoryUsage(model: ModelInfo): Long {
        // Base estimate: file size + 30% overhead for runtime
        val baseSize = model.memoryRequired ?: model.downloadSize ?: 0L
        return (baseSize * 1.3).toLong()
    }

    /**
     * Get optimal hardware configuration for a model
     */
    override fun optimalConfiguration(model: ModelInfo): HardwareConfiguration {
        return HardwareConfiguration(
            preferGPU = false, // ONNX uses CPU on mobile by default
            minMemoryMB = (estimateMemoryUsage(model) / 1024 / 1024).toInt(),
            recommendedThreads = 4 // Default thread count for cross-platform compatibility
        )
    }

    /**
     * Called when this adapter is registered with ModuleRegistry
     * Registers STT, TTS, and VAD providers
     */
    override fun onRegistration() {
        logger.info("ONNX adapter registration - registering service providers")

        // Register STT provider
        ModuleRegistry.shared.registerSTT(ONNXSTTServiceProvider())

        // Register TTS provider
        ModuleRegistry.shared.registerTTS(ONNXTTSServiceProvider())

        // Register VAD provider
        ModuleRegistry.shared.registerVAD(ONNXVADServiceProvider())

        logger.info("Registered ONNX STT, TTS, and VAD providers with ModuleRegistry")
    }

    /**
     * Get models provided by this adapter
     * Returns empty - models are discovered from external sources
     */
    override fun getProvidedModels(): List<ModelInfo> {
        return emptyList()
    }

    /**
     * Get download strategy for ONNX models
     */
    override fun getDownloadStrategy(): DownloadStrategy? {
        return ONNXDownloadStrategy()
    }

    /**
     * Get model storage strategy for ONNX models
     * Used to detect downloaded models on disk with ONNX-specific structures
     */
    override fun getModelStorageStrategy(): ModelStorageStrategy? {
        return ONNXModelStorageStrategy()
    }

    /**
     * Initialize a component with parameters
     */
    override suspend fun initializeComponent(
        parameters: ComponentInitParameters,
        modality: FrameworkModality
    ): Any? {
        val modelId = parameters.modelId ?: return createService(modality)

        // If model ID is provided, try to find and load it
        // In practice, the calling code should resolve the ModelInfo first
        logger.info("Initializing ONNX component for model: $modelId, modality: $modality")

        return createService(modality)
    }

    // MARK: - Private Methods

    /**
     * Cleanup cached services that haven't been used recently
     */
    private fun cleanupStaleCache() {
        val now = currentTimeMillis()

        lastSTTUsage?.let { lastUsage ->
            if (now - lastUsage > cacheTimeout) {
                logger.debug("Cleaning up stale STT service cache")
                cachedSTTService = null
                lastSTTUsage = null
            }
        }

        lastTTSUsage?.let { lastUsage ->
            if (now - lastUsage > cacheTimeout) {
                logger.debug("Cleaning up stale TTS service cache")
                cachedTTSService = null
                lastTTSUsage = null
            }
        }
    }

    companion object {
        /**
         * Singleton instance
         */
        val shared: ONNXAdapter = ONNXAdapter()

        /**
         * Register the ONNX adapter with ModuleRegistry
         * @param priority Registration priority (higher = selected first)
         */
        fun register(priority: Int = 100) {
            ModuleRegistry.shared.registerFrameworkAdapter(shared, priority)
        }
    }
}

// Platform-specific service creation (expect declarations)

/**
 * Create ONNX STT service from model path
 */
expect suspend fun createONNXSTTServiceFromPath(modelPath: String): Any

/**
 * Create ONNX TTS service from model path
 */
expect suspend fun createONNXTTSServiceFromPath(modelPath: String): Any
